release/models/bgp/openconfig-bgp-policy.yang

```diff
@@ -210,23 +217,37 @@ module openconfig-bgp-policy {
   typedef bgp-set-med-type {
     type union {
       type uint32;
-      type string {
-        pattern '[+-][0-9]+';
-        oc-ext:posix-pattern '^[+-][0-9]+$';
-      }
       type enumeration {
         enum IGP {
-          description "set the MED value to the IGP cost toward the
-          next hop for the route";
+          description
+            "Set the MED value to the IGP cost toward the next hop for
+            the route";
         }
       }
     }
     description
-      "Type definition for specifying how the BGP MED can
-      be set in BGP policy actions. The three choices are to set
-      the MED directly using the uint32 type or increment/decrement
-      using +/- notation in the string type or setting it to
-      the IGP cost using the enum (predefined value).";
+      "Type definition for specifying how the BGP MED can be set in BGP
+      policy actions.  The MED can be specified as a direct integer
+      value or setting it to the IGP cost.  When set to the integer
+      value without a MED 'action', the value will be explicitly set.
+      If a MED 'action' is defined, the previous value will be modified
+      according to the action (e.g. ADD or SUBTRACT).";
+  }
+
+  typedef bgp-set-med-action {
+    type enumeration {
+      enum ADD {
+        description
+          "Action to increment the previous MED value.";
+      }
+      enum SUBTRACT {
+        description
+          "Action to decrement the previous MED value.";
+      }
```

Comment:
I think what you have here works, but it would be more explicit if we add an enum `SET` or similar to indicate to change the value of med to be the integer specified (instead of add or subtract).  This way the bgp-set-med-action is always relevant when `set-med` is a uint32 value.

Reply:
We can very well enforce this new action leaf by doing such.  The current proposal was assuming if only `set-med` is set then it is an implicit "set" but if this leaf is specified in addition then the behavior changes to add/subtract.

If we add another enum variant here then it likely makes sense to gate a `when` statement as well otherwise we leave some undefined implementation behavior on the table

Reply:
After thinking about this a few days  I do prefer the extending the enum and as you suggest, adding a `when` constraint because the relationship between the enum and the leaf are explicit (and still rather simple) in yang .  

Would you like to make this change?

Reply:
Updated via commit: https://github.com/openconfig/public/pull/1165/commits/3d1aba4f520ddd6f8481f0751653a78837d5edc4 to add an explicit `SET` action along with constraints between the 2 leafs.

Example stub validation (removal of either leaf):
```
<bgp-actions xmlns="http://openconfig.net/yang/bgp-policy">
  <config>
    <set-med>100</set-med>
    <set-med-action>ADD</set-med-action>
  </config>
</bgp-actions>

libyang err : set-med-action cannot be specified without a set-med value (/openconfig-bgp-policy:bgp-actions/config/set-med-action)

libyang err : set-med cannot be specified without a valid set-med-action (/openconfig-bgp-policy:bgp-actions/config/set-med)
```

NOTE: When `IGP` is specified, this now enables the ability to apply any action in its current state which is likely fine

Reply:
Thanks for adding this!  

We reviewed again in Aug 27, 2024 operators meeting.  To make this backwards compatible as possible, the typedef bgp-set-med-action should have a default of 'SET'.     This way if someone is currently specifying an integer as the 'bgp-set-med-type', then we'll get the same behavior as today.

Reply:
I've been a little hesitant on the approach of using defaults in this fashion and would like to hear folks thoughts as this could be applied in a number of areas and there should likely be a bit of focus and guidance on default handling in general outside the scope of this PR.

On one hand, the approach suggested could provide a fraction of a transition path for a previous configuration.  The case where a precise positive value was specified only without prefix.  e.g. `200`

What would not transition whatsoever is anything that previously contained a `+/-` prefix even if the implementation skirted around this previously.

A few other thoughts

* If a default were set on `set-med-action` then there would no longer be any need for the `must` statements as `set-med-action` would always be present.  Since this value is always set, it can then exist w/o any value specified as well.
* When defaults are set in this fashion, we have an implicit behavior vs. enforcing explicit behavior.  The backwards incompatibilities introduced really needs to be understood and assessed imo - anything else may be luck with undesirable results.  A previous `+200` may have been treated one way previously but now still accepted w/ alternate behavior that needs explicit setting of the "action" node.
* Defaults in general tend to lack behavioral expectations with protocols such as gNMI (retrieval).  This is something already covered in NETCONF/RESTCONF.  A case of this is the expectations surrounding a "get" to the config branch where defaults are conveyed.  Today this should likely never return defaults but rather the mirrored state node would be emitted.

In addition to above regarding defaults is behavioral expectations surrounding what led to this PR.  If a value on a int type is set as `+/-` prefixed, the implementation really should report back any config/state as non `+` prefixed in the case of positive (dropping the prefix) or `-` prefixed (retained) in the case of negative.

Reply:
I am ok with leaving out a default.  When discussing with others offline I was also reminded of our bias to avoid defaults and prefer leaves being set explicitly by a caller.